niceman/formats/tests/test_niceman.py

```diff
@@ -36,6 +45,89 @@ def test_write():
+                    DEBPackage(
+                        name="libc-bin",
+                        upstream_name=None,
+                        version="2.24-11+deb9u3",
+                        architecture="amd64",
+                        source_name="glibc",
+                        source_version=None,
+                        size="779468",
+                        md5="3b9aaa83b5253895b8e13509659662e4",
+                        sha1=None,
+                        sha256="aaa",
+                        versions={"2.24-11+deb9u1": ["apt_Debian_stable_foo"],
+                                  "2.24-11+deb9u3": ["apt_Debian_stable_bar",
+                                                     "apt__now__0"]},
+                        install_date="2018-03-12 10:55:13+00:00",
+                        files=["/usr/bin/zdump"])],
+                version="9.4"),
+            CondaDistribution(
+                name="conda",
+                path="/path/to/miniconda3",
+                conda_version="4.4.10",
+                python_version="3.6.3.final.0",
+                platform="linux-64",
+                environments=[
+                    CondaEnvironment(
+                        name="root",
+                        path="/path/to/miniconda3",
+                        packages=[
+                            CondaPackage(name="condapkg"),
+                            CondaPackage(
+                                version="36.5.0",
+                                build="py36he42e2e1_0",
+                                name="setuptools",
+                                md5="cb1383539629db998105faf7e91e2bc7",
+                                url="https://somewhere")],
+                        channels=[
+                            CondaChannel(
+                                name="defaults",
+                                url="https://somewhere")]),
+                    CondaEnvironment(
+                        name="other",
+                        path="/path/to/miniconda3",
+                        packages=[
+                            CondaPackage(name="condapkg2")])]),
+            VenvDistribution(
+                name="venv0",
+                path="/usr/bin/virtualenv",
+                venv_version="15.1.0",
+                environments=[
+                    VenvEnvironment(
+                        path="venv-niceman",
+                        python_version="3.5.3",
+                        packages=[
+                            VenvPackage(
+                                version="3.12",
+                                name="PyYAML",
+                                location="/path/to/venv/site-packages",
+                                local=True)])]),
+            VenvDistribution(name="venv1")])
+
+    output = io.StringIO()
+    NicemanProvenance.write(output, spec)
+    loaded = NicemanProvenance(output.getvalue()).get_environment()
+    assert spec == loaded
```

Comment:
I wonder if we ever checked (have a test) that such comparisons of nested Attr classes actually work correctly (ie would fail if anything fails).  @ReproNim/trd3 anyone remembers?

Reply:
> I wonder if we ever checked (have a test) that such comparisons of nested Attr classes actually work correctly (ie would fail if anything fails).

It seems to as long as the default `attr.s(..., cmp=True)` is used.  `DEBPackage` and `DockerImage` are currently the only classes that set `cmp=False`.